File: bindings/go/ctlog.go
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

// CT Log protocol types for proven-servers.
// All tag values match the Idris2 ABI discriminants exactly.
package proven

// LogEntryType represents the LogEntryType type (Idris2 ABI tags).
type LogEntryType uint8

const (
	LogEntryTypeX509Entry LogEntryType = iota
	LogEntryTypePrecertEntry
)

// SignatureType represents the SignatureType type (Idris2 ABI tags).
type SignatureType uint8

const (
	SignatureTypeCertificateTimestamp SignatureType = iota
	SignatureTypeTreeHash
)

// MerkleLeafType represents the MerkleLeafType type (Idris2 ABI tags).
type MerkleLeafType uint8

const (
	MerkleLeafTypeTimestampedEntry MerkleLeafType = iota
)

// SubmissionStatus represents the SubmissionStatus type (Idris2 ABI tags).
type SubmissionStatus uint8

const (
	SubmissionStatusAccepted SubmissionStatus = iota
	SubmissionStatusDuplicate
	SubmissionStatusRateLimited
	SubmissionStatusRejected
	SubmissionStatusInvalidChain
	SubmissionStatusUnknownAnchor
)

// VerificationResult represents the VerificationResult type (Idris2 ABI tags).
type VerificationResult uint8

const (
	VerificationResultValidProof VerificationResult = iota
	VerificationResultInvalidProof
	VerificationResultInconsistentTree
	VerificationResultStaleSth
)

// ServerState represents the ServerState type (Idris2 ABI tags).
type ServerState uint8

const (
	ServerStateIdle ServerState = iota
	ServerStateActive
	ServerStateMerging
	ServerStateSigning
	ServerStateShutdown
)

File: bindings/ruby/lib/proven_servers/ctlog.rb
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# CT Log protocol types for proven-servers.

# frozen_string_literal: true

module ProvenServers
  # CT Log protocol types for proven-servers.
  module Ctlog
    # LogEntryType matching the Idris2 ABI tags.
    module LogEntryType
      X509_ENTRY = 0
      PRECERT_ENTRY = 1
    end

    # SignatureType matching the Idris2 ABI tags.
    module SignatureType
      CERTIFICATE_TIMESTAMP = 0
      TREE_HASH = 1
    end

    # MerkleLeafType matching the Idris2 ABI tags.
    module MerkleLeafType
      TIMESTAMPED_ENTRY = 0
    end

    # SubmissionStatus matching the Idris2 ABI tags.
    module SubmissionStatus
      ACCEPTED = 0
      DUPLICATE = 1
      RATE_LIMITED = 2
      REJECTED = 3
      INVALID_CHAIN = 4
      UNKNOWN_ANCHOR = 5
    end

    # VerificationResult matching the Idris2 ABI tags.
    module VerificationResult
      VALID_PROOF = 0
      INVALID_PROOF = 1
      INCONSISTENT_TREE = 2
      STALE_STH = 3
    end

    # ServerState matching the Idris2 ABI tags.
    module ServerState
      IDLE = 0
      ACTIVE = 1
      MERGING = 2
      SIGNING = 3
      SHUTDOWN = 4
    end

  end
end
